File: SparkStream/src/main/java/untis/AvroDeserializer.java
package untis;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.util.Map;

import org.apache.avro.io.BinaryDecoder;
import org.apache.avro.io.DatumReader;
import org.apache.avro.io.DecoderFactory;
import org.apache.avro.specific.SpecificDatumReader;
import org.apache.kafka.common.serialization.Deserializer;


/**
 * @Title AvroDeserializer.java
 * @Description 使用传统的 Avro API 自定义反序列类
 * @Author YangYunhe
 * @Date 2018-06-21 17:19:40
 */
//public class AvroDeserializer implements Deserializer<Stock> {
//
//    @Override
//    public void close() {}
//
//    @Override
//    public void configure(Map<String, ?> arg0, boolean arg1) {}
//
//    @Override
//    public Stock deserialize(String topic, byte[] data) {
//        if(data == null) {
//            return null;
//        }
//        Stock stock = new Stock();
//        ByteArrayInputStream in = new ByteArrayInputStream(data);
//        DatumReader<Stock> userDatumReader = new SpecificDatumReader<>(stock.getSchema());
//        BinaryDecoder decoder = DecoderFactory.get().directBinaryDecoder(in, null);
//        try {
//            stock = userDatumReader.read(null, decoder);
//        } catch (IOException e) {
//            e.printStackTrace();
//        }
//        return stock;
//    }
// }

File: SparkStream/src/main/scala/untis/AvroDeserializer.scala
package untis

import io.confluent.kafka.schemaregistry.client.SchemaRegistryClient
import io.confluent.kafka.serializers.AbstractKafkaAvroDeserializer
import org.apache.avro.generic.GenericRecord
import org.apache.kafka.common.serialization.StringDeserializer

class AvroDeserializer extends AbstractKafkaAvroDeserializer {
  def this(client: SchemaRegistryClient) {
    this()
    this.schemaRegistry = client
  }

  override def deserialize(bytes: Array[Byte]): String = {
    val value = super.deserialize(bytes)
    value match {
      case str: String =>
        str
      case _ =>
        val genericRecord = value.asInstanceOf[GenericRecord]
        if (genericRecord == null) {
          // 返回空字符串
          null
        } else {
          genericRecord.toString

        }
    }
  }
}
